shared/storage.go

```diff
@@ -284,3 +286,125 @@ func (cs byteCachedStore) Get(cacheID, storeID, iValue interface{}) error {
 func NewByteCachedStore(ctx context.Context, cache ReadWritable, store Readable) CachedStore {
 	return byteCachedStore{ctx, cache, store}
 }
+
+// ObjectStore is a store that populates an arbitrary output object on Get().
+type ObjectStore interface {
+	Get(id, value interface{}) error
+}
+
+// ObjectCache is an ObjectStore that also supports Put() for arbitrary id/value
+// pairs.
+type ObjectCache interface {
+	ObjectStore
+	Put(id, value interface{}) error
+}
+
+type jsonObjectCache struct {
+	ctx      context.Context
+	delegate ReadWritable
+}
+
+func (oc jsonObjectCache) Get(id, value interface{}) error {
+	r, err := oc.delegate.NewReadCloser(id)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err := r.Close()
+		if err != nil {
+			logger := oc.ctx.Value(DefaultLoggerCtxKey()).(Logger)
+			logger.Warningf("Error closing JSON object cache delegate ReadCloser: %v", err)
+		}
+	}()
+
+	data, err := ioutil.ReadAll(r)
+	if err != nil {
+		return err
+	}
+
+	return json.Unmarshal(data, value)
+}
+
+func (oc jsonObjectCache) Put(id, value interface{}) error {
+	w, err := oc.delegate.NewWriteCloser(id)
+	if err != nil {
+		return err
+	}
+
+	data, err := json.Marshal(value)
+	if err != nil {
+		return err
+	}
+
+	n, err := w.Write(data)
+	if err != nil {
+		return err
+	}
+	if n != len(data) {
+		return fmt.Errorf("JSON object cache: Attempted to write %d bytes, but wrote %d bytes", len(data), n)
+	}
+
+	err = w.Close()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// NewJSONObjectCache constructs a new JSON object cache, bound to the input
+// context.Context and delgating to the input ReadWritable.
+func NewJSONObjectCache(ctx context.Context, delegate ReadWritable) ObjectCache {
+	return jsonObjectCache{ctx, delegate}
+}
+
+type datastoreObjectStore struct {
+	ctx  context.Context
+	kind string
+}
+
+func (s datastoreObjectStore) Get(iID, value interface{}) error {
+	id, ok := iID.(int64)
```

Comment:
So, this is fine, but I question the need to abstract to non-int64-IDs. Do we ever identify something by a different ID type?

Reply:
Yes. Memcache uses strings.